### PR TITLE
[DSET-3928] [DSET-4094] Mid term approach for handling changed Kubernetes API URL endpoints in the Kubernetes monitors

### PR DESCRIPTION
**Pull request by tomaz-s1**
## Description

This pull request implements support for fallback URLs for various Kubernetes API endpoints.

Fallback URL(s) are used in case the primary / first URL returns 404 not found response.

This is meant as a simple mid term solution for a common use case we don't handle today - Kubernetes API endpoint has been promoted from beta to stable (aka URL endpoint changed from `v1beta` to `v1`).

It's in a no way full blown robust solution or a replacement for a proper Kubernetes API versioning and management in the agent - that's a much larger and longer term project.

The goal behind this proposed solution is to improve the status quo and solve the most pressing problem without introducing new technical debt or more complex edge cases. 

The solution could indeed be made complex (e.g. in-process caching of valid / working URLs, etc.), but that would make it more complicated (cache invalidation, thread safety, etc.) which means it would greatly increase the chance of bugs, edge cases, etc.

Yes, not caching the URL may have some overhead, but it's likely fine for now since it's not that common code path and only applies to ``CronJob`` resource type. And in case we do see it introduces tons of overhead, we can improve it incrementally over time.

To guard against the possible overhead or other issues, I also introduced a new config option / feature flag with which user can opt-out of this behavior in case it introduces problems / too much of a performance overhead / similar.

## Background, Context

Recently a user reported that they are missing some Kubernetes Events Monitor events which apply to ``CronJob`` resource type.

After digging through the code, changelog and trying to reproduce it locally, I was able to reproduce it and pin point it to the change in Kubernetes v1.25.0 - https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-25.

> CronJob
>
>The batch/v1beta1 API version of CronJob is no longer served as of v1.25.
>
 >   Migrate manifests and API clients to use the batch/v1 API version, available since v1.21.
>    All existing persisted objects are accessible via the new API
>    No notable changes

Technically, in this specific case, there is a workaround possible since events are still emitted for ``Job`` objects. But we still want to try to handle this breaking change and improve the status quo. And ideally also have a code pattern / starting point we can apply to similar problems in the future.

## Proposed Solution

I introduced a concept of "fallback" URLs. When querying Kubernetes API for specific resource information, we use templatized URL defined as constant in a Python module.

With this code change, we now support defining multiple URLs for each resource (e.g. ``CronJob``) and operation (single, list, list-all).

In case the first API endpoint returns 404 when querying for an object, we simply fall back to the next URL in the list and try querying that. We bail out on the first URL endpoint which doesn't return a 404 response. We intentionally only retry and try next URL on 404 and don't do it on any other error / status code.

I verified the change with older and newer Kubernetes versions and it works fine.

## TODO

- [x] Tests for all the code paths
- [x] Future flag which allows this functionality to be disabled (in case we uncover it adds a lot of overhead, unlikely, but still always a good idea to have a feature flag)
- [x] End to end tests (may be hard to cover both scenarios - negative and positive due to our end to end tests framework complexity, but at least I can add a test case which verifies ``CronJob`` event is present and that should be sufficient since we runs tests against multiple Kubernetes versions)
- [ ] ~~Refactor the code to reduce duplication~~

### CODE REVIEW

**Comment by tomaz-s1** [comment]
To make testing those changes easier and faster outside of docker / Kubernetes, I made a small change to the Kubernetes events monitor - I added a new `skip_leader_election` monitor config option.

When this option is set to `True` (defaults to `False`), leader election will be skipped and current process will be assumed to be a leader - 9b8f8b1048da9a06ba9efa94fc590b027d4e1eb4.

Keep in mind that this config option is only meant to be used for local development.

And even with that change in place, it's only meant to be used for initial quick local testing. Eventually and as part of CI/CD we of course still want to test everything end to end agent running on Kubernetes (which is indeed what we do on CI).

**Comment by tomaz-s1** [comment]
Pushed a small change / fix to `scripts/generate-docs-for-all-monitors.sh` script so it also works correctly on M1/M2 laptops where brew uses a different path for storing packages.

**Review by ArthurKamalov — APPROVED** [review]
LGTM

File: tests/e2e/k8s_events_monitor/cronjob_v1beta.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+apiVersion: batch/v1beta
+#apiVersion: batch/v2alpha1
```
[the remark this answers is not in the document]
Comment (tomaz-s1) [reply tests/e2e/k8s_events_monitor/cronjob_v1beta.yaml:2]: Yeah, it I left it there if we ever need  to test it with alpha, will remove it.

[the remark this answers is not in the document]
Comment (tomaz-s1) [reply tests/e2e/k8s_events_monitor/cronjob_v1beta.yaml]: Removed in eeb74ff22219af743c271e2153e88d33935bc057

**Pull request merged**